src/realm.js

```diff
@@ -128,6 +128,10 @@ export class Effects {
   canBeApplied: boolean;
   _id: number;
 
+  shallowCloneWithResult(result: Completion): Effects {
```

Comment:
Some comments around mutability would be nice.
1. I'd assume that an `Effects` instance is generally meant to be immutable. (With a twist regarding `canBeApplied`.)
2. However, this function indicates to me that it's main purposes is to allow for some mutability. Is this meant to be an "internal" API? However, it's difficult to tell, as I cannot see any immediate mutation after any call to `shallowCloneWithResult`.

Similarly for `Completion` and its `shallowCloneWithoutEffects`.

Reply:
I'm a bit at a loss as to your point. The clone methods do not make exact clones, but modify their results in the way that is spelled out in their names. As for the immutability of Effects, that is the goal but not yet the reality. When it becomes reality, there will still be lazy initialization. The canBeApplied flag is a temporary debugging aid.